src/frontend/src/optimizer/plan_node/batch_hash_agg.rs

```diff
@@ -79,7 +79,40 @@ impl ToDistributedBatch for BatchHashAgg {
             &Order::any(),
             &RequiredDist::shard_by_key(self.input().schema().len(), self.group_key()),
         )?;
-        Ok(self.clone_with_input(new_input).into())
+        if self.logical.can_two_phase_agg() && self.logical.two_phase_agg_forced() {
+            // partial agg
+            let partial_agg: PlanRef = self.clone_with_input(new_input).into();
+
+            // enforce exchange on input to partial agg.
+            // If we are going to shuffle, we should do it early,
+            // So we can aggregate more data in parallel.
+            let input_fields = self.base.schema().fields();
+            let input_col_num = input_fields.len();
+            let exchange = RequiredDist::shard_by_key(input_col_num, self.group_key())
+                .enforce_if_not_satisfies(partial_agg, &Order::any())?;
```

Comment:
This enforces exchange on input, rather than output of partial agg. This differs from two phase simple agg. I'm not sure if is the right behaviour.

Reply:
This exchange enforce the `partial_agg`, but it use `self.group_key()` as shuffle columns. I think we should use  `(0..self.group_key().len()).collect()` instead. BTW, batch simple agg can also be converted into 2 phase agg. Another advantage of two phase agg is that, it doesn't require to enforce any distribution to the partial agg's input, but we do need exchange between the partial agg and the global agg.

Reply:
> Another advantage of two phase agg is that, it doesn't require to enforce any distribution to the partial agg's input, but we do need exchange between the partial agg and the global agg.

Why do we need exchange if distribution is the same between partial and global agg?

Reply:
If distribution is the same between partial and global agg, I think we don't even need 2 phase agg, because 1 phase is enough.